### ISSUE TITLE
Your Vector lib is simply unusable. 

### ISSUE THREAD

**Issue by SandvichDISH** [INPStarfall/Starfall#230]
First of all, there is no division function, so that's unbelievably impractical.
Second, the fact that you can't multiply a Vector by another Vector is also unbelievably impractical. It's not the worst thing in the world, but considering I had to use Vectors as Colors because someone decided you can't do arithmetic on Colors, this now makes doing mathematics with colours even more unbearable. 
Then, sending a Vector from the server to the client through the net lib with writeTable and readTable (as I do in my project) somehow CONVERTS the sent Vector from a Starfall Vector to a Garry's Mod Vector. For example:

``` lua
--@name 
--@author 
--@sharedscreen

if SERVER then
    timer.simple( 1, function()
        if net.canSend() then
            net.start( "test" )
                net.writeTable( { Vector( 1, 1, 1 ) } )
            net.send()
        end
    end )
else 
    local testvar1 = nil
    local testvar2 = Vector( 1, 1, 1 )

    hook.add( "net", "testnet", function( msg ) 
        if msg == "test" then
            testvar1 = net.readTable()[ 1 ]
            print( testvar1, type( testvar1 ), testvar2, type( testvar2 ) )
            print( testvar1:GetNormalized() ) -- What the hell
            print( testvar2:getNormalized() ) -- is this
            print( testvar1 + testvar2 ) -- this obviously doesn't work then
        end
    end ) 
end
```

Aside from that I think it's pretty silly that even when you replace Garry's library (which is perfectly functional anyway) you still use Vector( x, y, z ) instead of vector( x, y, z ) despite the fact that you use camelCasing for everything else you do.

**Comment by awilliamson** [comment]
Vector-Vector division, mathematically, doesn't exist. There isn't a single definition of vector-vector multiplication as you have cross product and dot product. Since division is the inverse operation of multiplication, you can't have single division function for vector-vector.

Vector-Scalar multiplication exists, it uses the __mul metamethod. So v \* s would work just fine. If you want vector-vector use the methods provided.

Colors, in Garry's Mod, are just a table with three values in it. If you can give examples of what you'd want to do with Color arithmetic then go ahead. If you want convenience functions then make them yourself and use them. It's not difficult for colors as you have the r, g, and b properties and I believe numerical indexing. Similar to Vectors, x, y, z and 1, 2, 3.

camelCasing is for when you have two words for a variable, if there is only one word it should be lowercase. The reason Vector is uppercase is because it's a constructor it makes a new vector for you ( with the specified params ) and returns it.

The reason we replace Garry's tacked on lua is because what he adds can be implemented in pure lua anyway. They function but some of them don't make perfect sense. There exists duplicates of functions, they're naming is not exactly correct. And we're hoping Starfall will go on to teach people Lua and skills related to computer science that they can use in the future.

I will look into the net issue to see what is going on, thanks for raising that point.

**Comment by Xandaros** [comment]
Dividing vectors by scalars is planned, see #216 
Multiplying vectors by vectors _is_ possible, see [Documentation](http://sf.inp.io/classes/Vector.html), we provide the methods dot, cross and mul. We are even nice enough to provide cross-product, which is a corner-case, but happens to be insanely useful in gmod.

Arithmetic on colours? What's cyan+red? Or blue*5?

The net thing is a legitimate bug, @awilliamson is working on it.

Methods and functions are in camelCase, yes. But constructors are not. Notice how all constructors are uppercased? (Vector, Angle, Color, ...)

**Comment by awilliamson** [comment]
Just tested it with Color, they have the same problem with being converted to Vector. This is an issue with our custom types as our Sanitize/UnSanitize functions don't handle them correctly. We use those when converting all types within tables for sending.

**Comment by SandvichDISH** [comment]
So, I just accidentally removed my comment. GG me. Let me write that again:

When I talk about dividing or multiplying the Vector, I simply mean something akin to

``` lua
newVector = Vector( 10, 10, 10 ) / 10
--newVector == Vector( 1, 1, 1 )
newVector = Vector( 10, 10, 10 ) / Vector( 10, 5, 1 )
--newVector == Vector( 1, 2, 10 )
newVector = Vector( 1, 1, 1 ) * 10
--newVector == Vector( 10, 10, 10 )
newVector = Vector( 1, 1, 1 ) * Vector( 10, 0, 20 )
--newVector == Vector( 10, 0, 20 )

-- In short:
function Vector.__mul( a, b ) return type( b ) == "number" and Vector( a.x * b, a.y * b, a.z * b ) or Vector( a.x * b.x, a.y * b.y, a.z * b.z ) end -- Or something 
```

And have something identical to that for colours. Maybe it doesn't make sense to you, but last time I checked when I code a shader in GLSL (or an addon for Garry's Mod) basic functionality like that is provided and incredibly useful.

I think this isn't too much to ask since in Garry's Mod traditionally, vectors have never been only vectors in the mathematical sense. Vectors are also used as coordinates, for example. So when I want to shrink down this coordinate, let's say "camera position" to a 16th of it's original size (for raytracing inside the skybox) I would go

``` lua
coordinate = coordinate / 16 -- coordinate being a world position
```

Again, I don't find this wholly unreasonable. Especially considering functions like trace.trace( startPosition, endPosition ) use **vectors** (despite the name) for the start and end **position**. 

You can even rename vectors to "number3" (HLSL style) or something if you absolutely insist on using mathematical terms, you can have unsoiled beautiful vectors in your own code, neatly separated from the rest, without taking away some very useful functionality that both Garry's Mod itself and E2 do have.

**Comment by Xandaros** [comment]
> newVector = Vector( 10, 10, 10 ) / 10

planned

> newVector = Vector( 10, 10, 10 ) / Vector( 10, 5, 1 )

Doesn't make sense

> newVector = Vector( 1, 1, 1 ) \* 10

Already implemented

> newVector = Vector( 1, 1, 1 ) \* Vector( 10, 0, 20 )

Ambiguous. Could be cross or dot product. (Although usually you mean dot product)
Component-wise multiplication is usually not defined and you have to do it yourself. If you need a function to do that, then make it yourself. It isn't too much to ask, is it?

> vectors have never been only vectors in the mathematical sense. Vectors are also used as coordinates

If you use vectors for a coordinate, that is called a position vector and only means that the vector begins in the origin. (As opposed to wherever you like, as with "normal" vectors. The only difference is how these vectors are interpreted, there is technically no difference.)

Vectors have ALWAYS been vectors in the mathematical sense. Going away from mathematics is what wouldn't make sense.

> coordinate = coordinate / 16

Again, this is planned. Did you even look at the issue I linked?

> You can even rename vectors to "number3"

They are called vectors, because they are vectors. "number3" implies a 3-tuple of numbers, which would leave you with even less operations than you get with vectors. There is a reason a vector is not defined by its components, but rather by its direction, orientation and length. It just so happens that specifying the components is a good representation, they aren't that important for the vector itself.

**Comment by SandvichDISH** [comment]
Well if you want to be pedantic about it, you can remove this functionality that Garry's library has. I'm no computer scientist or mathematician, but I have enjoyed Garry's Mod a lot over the past 6 years, and quite frankly if you're going to replace Garry's libraries and reduce the functionality in return over the sake of semantics then be my guest. But I will stop using starfall. 

Because unlike you, Garry understands that being able to do ~~"Color( 255, 255, 255 ) \* Color( 0, 1, 0 )"~~ "Vector( 255, 255, 255 ) \* Vector( 0, 1, 0 )" is INCREDIBLY useful and you're doing nothing but being annoying by removing that functionality and forcing me to create a function like colorMultiply( Color, Color ). You can feel all "proper" about it, saying that it doesn't make sense to multiply colours or that calling something "number3" would give even less operations, or you could just add all the operations you could ever want and have an amazing amount of functionality (as my HLSL example would imply).  I would gladly make my own colour "class" so I can have the functionality instead but then I have to create functions to convert types in a roundabout stupid way too, like render.setColor( toColor( colour( 255, 0, 0 ) \* 0.5 ) ) for example.

**TL;DR**: Be pedantic if you want, ant-fuck about semantics if you want, but you're not going to improve the experience of working with your addon in doing so.

**Comment by Xandaros** [comment]
> Garry understands that being able to do "Color( 255, 255, 255 ) \* Color( 0, 255, 0 )" is INCREDIBLY useful

![](http://i.imgur.com/1BV0QGE.png)

I don't even know what that is supposed to return. 0, 65025, 0? That's not a valid colour, though...

> saying that it doesn't make sense to multiply numbers

It does! But colours, doesn't. If you can tell me what cyan*blue is and why that makes sense, I will implement it, but at the moment, I don't see how multiplying colours would be possible.

> then I have to create functions to convert types in a roundabout stupid way too, like render.setColor( toColor( colour( 255, 0, 0 ) \* 0.5 ) ) for example

I thought about this a few days ago and I plan to improve this, but this has a rather low priority. All these functions you are suggesting don't really make sense, except for maybe component-wise vector multiplication. But that is a single function, you don't need to make your own vectors for that.

> TL;DR: Be pedantic if you want, ant-fuck about semantics if you want, but you're not going to improve the experience of working with your addon in doing so.

Nobody is forcing you to use it.

Edit: Maybe some clarification on why we are so opposed to adding things that don't make perfect sense: Starfall is great because it is pure and not full of garbage like E2 is. Sure, it makes E2 easy to use, but I prefer the [KISS-principle](http://en.wikipedia.org/wiki/KISS_principle), which we adopt in starfall. We don't define every single corner-case there is, but rather give you the opportunity to design your environment the way you want and keep the global namespace decluttered.

**Comment by SandvichDISH** [comment]
I edited my comment to fix some mistakes I made in writing it, which you might want to look at. I am wrong that you can multiply colors in gmod lua, apparently. I'm too used to that being commonplace functionality though when working with stuff like GLSL and HLSL.

**Comment by awilliamson** [comment]
>  I'm no computer scientist or mathematician, but I have enjoyed Garry's Mod a lot over the past 6 years, and quite frankly if you're going to replace Garry's libraries and reduce the functionality in return over the sake of semantics then be my guest.

You claim to have no formal or experience based Computer Science knowledge, yet you are quite adamant that what we're doing is 'wrong' and you're actually willing to stop using it because of this. As @Xandaros mentioned, we're not forcing you to use it. We're trying to help the community by providing something that is as correct as we can make it so that people don't learn incorrectly, you're an excellent example to prove our case.

I'd just like you to know that Garry had no formal education in Computer Science which is why some of the things he's implemented are bad for people just getting into programming. It's why we filter some of his things out because it is in the mix of things when it shouldn't and is sometimes not optimally implemented.

**Comment by SandvichDISH** [comment]
I'm not claiming anything you do is wrong, I'm claiming what you do is very impractical in the context of Garry's Mod.

**Comment by awilliamson** [comment]
> I'm not claiming anything you do is wrong, I'm claiming what you do is very impractical in the context of Garry's Mod.

Starfall is Lua sandbox, regardless of context. It just provides interfacing with Garry's Mod through the use of exposed Lua functions from GMod. It can relatively easily be transferred to any game which has Lua as it's embedded language. The core behind Starfall is pure Lua, the only thing that is GLua are the libraries which we use to interface with GMod.

Lua is an actual language which is decided upon by a committee, these are usually academics who discuss the additions or removal of features. Garry, isn't formally trained, and he added the features to GLua himself without consultation. What we're simply doing, is looking at the functions he's providing and making a decision about them as a group, whether they're good or bad and whether they should be included or not. We decided to omit the majority of Garry functions because most of them don't cross the C boundary, therefore can be implemented by users and sometimes better than what is provided. In the case of the Vector \* Vector example, people like you assume it's mathematically defined and argue for it's inclusion without understanding the underlying principles of the data type.

From pure mathematics, if you see V . V ( standard notation for multiplication in mathematics ) you will assume dot product between the two. With v \* v in GMod, you would therefore assume dot product or the unlikely case of cross product. However Garry, in his infinite wisdom, chose to implement it such that it multiplied the vectors component-wise which, if you know vectors, is completely wrong. There is a definition of component-wise multiplication called the Hadamard product or the entrywise product, this is defined for matrices, where vectors are a special case of matrices where the matrix is N by 1, where N is the number of dimensions.

To show you what I mean consider the following two links:
http://www.wolframalpha.com/input/?i=%7B1%2C2%2C3%2C7%7D*%7B4%2C5%2C6%2C8%7D
http://www.wolframalpha.com/input/?i=%7B1%2C2%2C3%7D*%7B4%2C5%2C6%7D

One of them shows cross-product the other shows dot product. But they are both using the same syntax, none of which demonstrates component-wise multiplication.

And for the curiosity of @Xandaros, can you define what cyan \* blue is?

**Comment by awilliamson** [comment]
Going to close as the only issue that was raised here is now fixed and it seems that @SandvichDISH has nothing further to add.
